### PR TITLE
fix(cloud_firestore): allow negative timestamp values

### LINKED ISSUE #3545: Timestamp.fromMillisecondsSinceEpoch() throws when calculating negative milliseconds.
Currently, this will throw an error:

```dart
Timestamp.fromMillisecondsSinceEpoch(-1);
```

I suggest the above code should be add as a test to the package.

It's a simple fix. The seconds/nanoseconds calculation is wrong:

```dart
    int milliseconds = -1;
    var _kThousand = 1000;
    var _kMillion = 1000000;

    // What is should be:
    int seconds = (milliseconds / _kThousand).floor();
    int nanoseconds = (milliseconds - seconds * _kThousand) * _kMillion;
    print('seconds = $seconds');
    print('nanoseconds = $nanoseconds');

    // Flutterfire's current code:
    seconds = (milliseconds ~/ _kThousand).floor();
    nanoseconds = (milliseconds - seconds * _kThousand) * _kMillion;
    print('seconds = $seconds');
    print('nanoseconds = $nanoseconds');
```

This is the same as https://github.com/flutter/flutter/issues/34929 and https://github.com/FirebaseExtended/flutterfire/issues/1222 which was not fixed after all.

This is a serious issue with an easy fix. Please, give it a high priority and don't let it stand for long. Thanks.

### PR DESCRIPTION
**Pull request by dackers86**
## Description

Negative values would throw an error with a value less than 0. This now allows for positive and negative values within a valid range.

## Related Issues

Fixes #3545

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`).
This will ensure a smooth and quick review process. Updating the `pubspec.yaml` and changelogs is not required.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors (See [Contributor Guide]).
- [x] All existing and new tests are passing.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] The analyzer (`flutter analyze`) does not report any problems on my PR.
- [x] I read and followed the [Flutter Style Guide].
- [x] I signed the [CLA].
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (please indicate a breaking change in CHANGELOG.md and increment major revision).
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/FirebaseExtended/flutterfire/blob/master/CONTRIBUTING.md
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[CLA]: https://cla.developers.google.com/

### CODE REVIEW

File: packages/cloud_firestore/cloud_firestore_platform_interface/test/timestamp_test.dart [hunk 1]
```diff
@@ -34,5 +35,49 @@ void main() {
...
+    });
+
+    test('fromMillisecondsSinceEpoch can handle future date', () {
+      int currentEpoch = DateTime.now().millisecondsSinceEpoch;
+      Timestamp t = Timestamp.fromMillisecondsSinceEpoch(currentEpoch);
+
+      expect(t.toDate().year > 1970, equals(true));
+    });
+
+    test('fromMillisecondsSinceEpoch can handle future date', () {
```
Comment (Ehesp) [inline packages/cloud_firestore/cloud_firestore_platform_interface/test/timestamp_test.dart:51]: Duplicate test name here

**Review by Ehesp — CHANGES_REQUESTED** [review]

File: packages/cloud_firestore/cloud_firestore_platform_interface/test/timestamp_test.dart (hunk 1 above)
Comment (dackers86) [reply packages/cloud_firestore/cloud_firestore_platform_interface/test/timestamp_test.dart:51]: Thanks. Updated!

File: packages/cloud_firestore/cloud_firestore_platform_interface/test/timestamp_test.dart [hunk 2]
```diff
@@ -36,5 +37,49 @@ void main() {
...
+    });
+
+    test('fromMillisecondsSinceEpoch can handle 0', () {
+      Timestamp t = Timestamp.fromMillisecondsSinceEpoch(0);
+      expect(t.toDate().year, 1970);
+      expect(t.toDate().month, 1);
+      expect(t.toDate().day, 1);
+    });
+
+    test('fromMillisecondsSinceEpoch can handle nagtive millisecond values',
```
Comment (mikehardy) [inline packages/cloud_firestore/cloud_firestore_platform_interface/test/timestamp_test.dart:70]:
```suggestion
    test('fromMillisecondsSinceEpoch can handle negative millisecond values',
```

just spelling

**Review by mikehardy — APPROVED** [review]
LGTM - the related issue was quite informative

**Comment by mikehardy** [comment]
I don't the CLA bot is about me, I have one on file and have successfully submitted PRs here :thinking:

**Comment by Salakar** [comment]
@dackers86 can you double check you've signed the CLA again please, it might be due to a changed GitHub email or something.

**Comment by kroikie** [comment]
@mikehardy could you try signing the CLA once more?

**Comment by mikehardy** [comment]
@googlebot I fixed it.

(apparently github@mikehardy.net was not one of my alternate emails in my google organization for this user, but (weirdly!) I have successfully merged PRs here before. Go figure :-) )

**Comment by mikehardy** [comment]
@kroikie CLA all set now, I'm confused how it happened but I was able to fix. Onwards!

**Pull request merged**